File: solutions/haskell/acronym/1/src/Acronym.hs
module Acronym (abbreviate) where

import Data.Char (toUpper, isUpper)

abbreviate :: String -> String
abbreviate s = concat (map initials (tokenize s))
  where
    initials :: String -> String
    initials s
      | allUpper == []                  = [toUpper (head s)]
      | (length allUpper) == 1          = allUpper
      | (length allUpper) == (length s) = [head s]
      | otherwise                       = allUpper
      where
        allUpper = (filter isUpper s)
    tokenize :: String -> [String]
    tokenize s2 = words (replaceAll s2)
    replaceAll :: String -> String
    replaceAll xs = map replace xs
    replace :: Char -> Char
    replace x = case x of
                  '-' -> ' '
                  c -> c

File: solutions/erlang/acronym/1/src/acronym.erl
-module(acronym).
-import(string, [tokens/2, to_upper/1]).
-import(lists, [map/2]).
-export([abbreviate/1]).

head([])    -> {error, "head on empty list"};
head([H|_]) -> H.

abbreviate(Phrase) ->
  to_upper(
    map(
      fun head/1,
      tokens(Phrase, " -_"))).
